Improve getting started information in order to avoid issues as 32 (#46)
operator-framework/getting-started: README.md

  - [Create a new project](#create-a-new-project)
  - [Manager](#manager)
- [Add a new Custom Resource Definition](#add-a-new-custom-resource-definition)
  - [Define the Memcached spec and status](#define-the-memcached-spec-and-status)
- [Add a new Controller](#add-a-new-controller)
  - [Resources watched by the Controller](#resources-watched-by-the-controller)
  - [Reconcile loop](#reconcile-loop)
- [Build and run the operator](#build-and-run-the-operator)
  - [1. Run as a Deployment inside the cluster](#1-run-as-a-deployment-inside-the-cluster)
  - [2. Run locally outside the cluster](#2-run-locally-outside-the-cluster)
- [Create a Memcached CR](#create-a-memcached-cr)
  - [Update the size](#update-the-size)
  - [Cleanup](#cleanup)
- [Reference implementation](#reference-implementation)
- [Manage the operator using the Operator Lifecycle Manager](#manage-the-operator-using-the-operator-lifecycle-manager)
  - [Generate an operator manifest](#generate-an-operator-manifest)
  - [Deploy the Operator](#deploy-the-operator)
  - [Create an application instance](#create-an-application-instance)
  - [Update an application](#update-an-application)
- [Conclusion](#conclusion)

> <!-- END doctoc generated TOC please keep comment here to allow auto update -->

Reviewer: :+1: I like the TOC.

### Create a new project

1. Use the CLI to create a new `memcached-operator` project:

```sh
$ mkdir -p $GOPATH/src/github.com/example-inc/
$ cd $GOPATH/src/github.com/example-inc/
$ operator-sdk new memcached-operator
$ cd memcached-operator
```

This creates the `memcached-operator` project.

> 2. Install dependencies by running `go mod tidy` 

Author: 👍

> **NOTE**: Learn more about the project directory structure from the SDK [project layout][layout_doc] documentation.

Reviewer 2: ```suggestion
**NOTE:** Learn more about the project directory structure from the SDK [project layout][layout_doc] documentation.
```

Author: 👍

$ operator-sdk generate k8s
```

> Also run the following command in order to automatically generate the OpenAPI validations. 

Author: 👍

```sh

> operator-sdk generate openapi  

Author: 👍

```

> You can check the changes applied in the `deploy/crds/cache_v1alpha1_memcached_crd.yaml`

Reviewer 2: ```suggestion
You can see the changes applied in `deploy/crds/cache_v1alpha1_memcached_crd.yaml`
```

Author: 👍

## Build and run the operator

> 

Author: 👍

Before running the operator, the CRD must be registered with the Kubernetes apiserver:

```sh

> $ kubectl create -f deploy/crds/cache_v1alpha1_memcached_crd.yaml 

Author: 👍

$ sed -i 's|REPLACE_IMAGE|quay.io/<user>/memcached-operator:v0.0.1|g' deploy/operator.yaml
# On OSX use:
$ sed -i "" 's|REPLACE_IMAGE|quay.io/<user>/memcached-operator:v0.0.1|g' deploy/operator.yaml
$ docker push quay.io/<user>/memcached-operator:v0.0.1
```

The above command will replace the string `REPLACE_IMAGE` with the <image>:<tag> build and pushed to your registry then verify that your `operator.yaml` file was updated successfully. Following an example.

```yaml
serviceAccountName: memcached-operator
containers:
- name: memcached-operator
  # Replace this with the built image name
  image: quay.io/example/memcached-operator:v0.0.1
  command:
  - memcached-operator
  imagePullPolicy: Always
```

**IMPORTANT:** Ensure that your cluster is able to pull the image pushed to your registry.

The Deployment manifest is generated at `deploy/operator.yaml`. Be sure to update the deployment image as shown above since the default is just a placeholder.

Setup RBAC and deploy the memcached-operator:

```sh
$ kubectl create -f deploy/service_account.yaml 
$ kubectl create -f deploy/role.yaml
$ kubectl create -f deploy/role_binding.yaml
$ kubectl create -f deploy/operator.yaml
```

> **NOTE:** To apply the RBAC you need be logged in as a user with cluster privileges, for example with the system:admin user to test it. (E.g. By using: `oc login -u system:admin.`) 

Reviewer: To apply the RBAC you need to be logged in

Reviewer: add tick marks around system:admin -> `system:admin`

Author: Makes sense keep it simple. Tks.

Verify that the memcached-operator deployment is up and running in the `memcached` namespace :

```sh
$ kubectl get deployment
NAME                     DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
memcached-operator       1         1         1            1           1m
```

> Verify that the memcached-operator pod is up and running:

Reviewer: I would put tick marks around memcached-operator -> `memcached-operator`

Author: Goo catcher. Added in all places that it was missing.

```sh
$ kubectl get pod
NAME                                  READY     STATUS    RESTARTS   AGE
memcached-operator-7d76948766-nrcp7   1/1       Running   0          44s
```

**IMPORTANT:** Ensure that you built and pushed the image, and updated the `operator.yaml` file.   

> Verify that the operator is running successfully by checking it logs. 

Reviewer: typo: it logs -> its logs

Author: 👍

{"level":"info","ts":1567613604.0976534,"logger":"cmd","msg":"Could not create ServiceMonitor object","error":"no ServiceMonitor registered with the API"}
{"level":"info","ts":1567613604.0986068,"logger":"cmd","msg":"Install prometheus-operator in your cluster to create ServiceMonitor objects","error":"no ServiceMonitor registered with the API"}
{"level":"info","ts":1567613604.0988326,"logger":"cmd","msg":"Starting the Cmd."}
{"level":"info","ts":1567613604.2993534,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"memcached-controller"}
{"level":"info","ts":1567613604.3995395,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"memcached-controller","worker count":1}
```

The following error will be faced if your cluster was unable to pull the image. 

**NOTE:** Just for tests purposes make the image public and setting up the cluster to allow use insecure registry. ( E.g --insecure-registry 172.30.0.0/16 )  

```sh
$ kubectl get pod
NAME                                  READY     STATUS             RESTARTS   AGE

> memcached-operator-6b5dc697fb-t62cv   0/1       ImagePullBackOff   0          2m

Reviewer: I would move this entire block just below the "The follo9wing error will occur ..." BEFORE the **NOTE:** line. 

It should read as follows:

The following error will occur if your cluster was unable to pull the image:
```sh
$ kubectl get pod
NAME                                  READY     STATUS             RESTARTS   AGE
memcached-operator-6b5dc697fb-t62cv   0/1       ImagePullBackOff   0          2m
```
**NOTE:** For testing purposes you can fix this error by making the image public and setting up your cluster to use an insecure image registry. To do so with docker, run the daemon with `--insecure-registry 172.30.0.0/16`.

Reviewer: I like to see the error you were pointing out immediately after you mentioned it. The NOTE is a reminder to me on how I should fix the above error.

Author: done 👍

### 1. Run as a Deployment inside the cluster

> Build the memcached-operator image and push it to your registry as the following example using https://quay.io.

Author: Done

> **NOTE**: You are able to you the https://hub.docker.com/ if you wish. 

Author: Oks, removed

```sh
$ operator-sdk build quay.io/<user>/memcached-operator:v0.0.1
$ sed -i 's|REPLACE_IMAGE|quay.io/<user>/memcached-operator:v0.0.1|g' deploy/operator.yaml
# On OSX use:
$ sed -i "" 's|REPLACE_IMAGE|quay.io/<user>/memcached-operator:v0.0.1|g' deploy/operator.yaml
$ docker push quay.io/<user>/memcached-operator:v0.0.1
```

> The above command will replace the string `REPLACE_IMAGE` with the <image>:<tag> build and pushed to your registry then verify that your `operator.yaml` file was updated successfully. Following an example.

Author: Done 👍

```yaml
serviceAccountName: memcached-operator
containers:
- name: memcached-operator
  # Replace this with the built image name

>   image: quay.io/example/memcached-operator:v0.0.1

Author: Great spot. Done 👍

  command:
  - memcached-operator

>   imagePullPolicy: Always 

Reviewer 2: ```suggestion
  imagePullPolicy: Always
```

Author: 👍

```

> **IMPORTANT:** Ensure that your cluster is able to pull the image pushed to your registry. 

Reviewer 2: ```suggestion
**IMPORTANT:** Ensure that your cluster is able to pull the image pushed to your registry.
```

Author: 👍

The Deployment manifest is generated at `deploy/operator.yaml`. Be sure to update the deployment image as shown above since the default is just a placeholder.

Setup RBAC and deploy the memcached-operator:

```sh

> $ kubectl create -f deploy/service_account.yaml 

Author: Done 👍

$ kubectl create -f deploy/role.yaml
$ kubectl create -f deploy/role_binding.yaml
$ kubectl create -f deploy/operator.yaml
```

**NOTE**:To apply the RBAC you need be logged in as a user with cluster privileges, for example with the system:admin user to test it. (E.g. By using: `oc login -u system:admin.`) 

> Verify that the memcached-operator deployment is up and running in the `memcached` namespace :

Author: Done and tks.

```sh
$ kubectl get deployment
NAME                     DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   AGE
memcached-operator       1         1         1            1           1m
```

Verify that the memcached-operator pod is up and running:

```sh
$ kubectl get pod
NAME                                  READY     STATUS    RESTARTS   AGE
memcached-operator-7d76948766-nrcp7   1/1       Running   0          44s
```

> **IMPORTANT:** Please, ensure that you built and pushed the image as updated the `operator.yaml file with.   

Author: 👍

{"level":"info","ts":1567613603.9398162,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"memcached-controller","source":"kind source: /, Kind="}
{"level":"info","ts":1567613604.0514739,"logger":"metrics","msg":"Metrics Service object created","Service.Name":"memcached-operator-metrics","Service.Namespace":"myproject"}
{"level":"info","ts":1567613604.0976534,"logger":"cmd","msg":"Could not create ServiceMonitor object","error":"no ServiceMonitor registered with the API"}
{"level":"info","ts":1567613604.0986068,"logger":"cmd","msg":"Install prometheus-operator in your cluster to create ServiceMonitor objects","error":"no ServiceMonitor registered with the API"}
{"level":"info","ts":1567613604.0988326,"logger":"cmd","msg":"Starting the Cmd."}
{"level":"info","ts":1567613604.2993534,"logger":"kubebuilder.controller","msg":"Starting Controller","controller":"memcached-controller"}
{"level":"info","ts":1567613604.3995395,"logger":"kubebuilder.controller","msg":"Starting workers","controller":"memcached-controller","worker count":1}
```

> The following error will be faced if your cluster was unable to pull the image. 

Author: Done 👍

> **NOTE:** Just for tests purposes make the image public and setting up the cluster to allow use insecure registry. ( E.g --insecure-registry 172.30.0.0/16 )  

Reviewer 2: ```suggestion
**NOTE:** For testing purposes you can fix this error by making the image public and setting up your cluster to use an insecure image registry. To do so with docker, run the daemon with `--insecure-registry 172.30.0.0/16`.
```

Author: Done 👍

```sh
$ kubectl get pod
NAME                                  READY     STATUS             RESTARTS   AGE
memcached-operator-6b5dc697fb-t62cv   0/1       ImagePullBackOff   0          2m
```

> Following the logs in the error scenario described above. 

Reviewer 2: ```suggestion
```

Reviewer: I would keep the line but rewrite it. Something like this:

The error mentioned above can also be seen in the operator's logs:
```
show command how to get logs and show the logs here
```

Author: I liked your suggestion @jmrodri 👍 .I think is important shows the error in order to avoid the users raise issues to understand what is happen since it is a common issue scenario.

```sh

> $kubectl logs memcached-operator-6b5dc697fb-t62cv

Reviewer 2: ```suggestion
$ kubectl logs memcached-operator-6b5dc697fb-t62cv
```

Author: Done 👍

Delete the operator and its related resources:

```sh

> $ kubectl delete -f deploy/crds/cache_v1alpha1_memcached_crd.yaml 

Reviewer 2: ```suggestion
```

This is run below.

Author: 👍
